File: sample/src/main/java/contacts/sample/cheatsheet/customdata/java/IntegrateGenderCustomDataActivity.java
package contacts.sample.cheatsheet.customdata.java;

import static contacts.core.WhereKt.isNotNull;

import android.app.Activity;

import java.util.List;

import contacts.core.*;
import contacts.core.data.*;
import contacts.core.entities.*;
import contacts.core.entities.custom.CustomDataRegistry;
import contacts.entities.custom.gender.*;

public class IntegrateGenderCustomDataActivity extends Activity {

    Contacts contacts = ContactsFactory.create(
            this, false, new CustomDataRegistry().register(new GenderRegistration())
    );

    List<Contact> getContactsWithGenderCustomData() {
        return contacts
                .query()
                .where(isNotNull(GenderFields.Type))
                .find();
    }

    Insert.Result insertRawContactWithGenderCustomData() {
        NewGender newGender = new NewGender(GenderEntity.Type.MALE);

        NewRawContact newRawContact = new NewRawContact();
        RawContactGenderKt.setGender(newRawContact, contacts, newGender);

        return contacts
                .insert()
                .rawContacts(newRawContact)
                .commit();
    }

    Update.Result updateRawContactGenderCustomData(RawContact rawContact) {
        MutableRawContact mutableRawContact = rawContact.mutableCopy();
        MutableGenderEntity mutableGender = RawContactGenderKt.gender(mutableRawContact, contacts);
        if (mutableGender != null) {
            mutableGender.setType(GenderEntity.Type.FEMALE);
        }

        return contacts
                .update()
                .rawContacts(mutableRawContact)
                .commit();
    }

    Update.Result deleteGenderCustomDataFromRawContact(RawContact rawContact) {
        MutableRawContact mutableRawContact = rawContact.mutableCopy();
        RawContactGenderKt.setGender(mutableRawContact, contacts, (MutableGenderEntity) null);

        return contacts
                .update()
                .rawContacts(mutableRawContact)
                .commit();
    }

    List<Gender> getAllGenders() {
        return GenderDataQueryKt.genders(contacts.data().query()).find();
    }

    DataUpdate.Result updateGender(MutableGender gender) {
        return contacts.data().update().data(gender).commit();
    }

    DataDelete.Result deleteGender(Gender gender) {
        return contacts.data().delete().data(gender).commit();
    }
}

File: sample/src/main/java/contacts/sample/cheatsheet/customdata/kotlin/IntegrateGenderCustomDataActivity.kt
package contacts.sample.cheatsheet.customdata.kotlin

import android.app.Activity
import contacts.core.*
import contacts.core.data.*
import contacts.core.entities.*
import contacts.core.entities.custom.CustomDataRegistry
import contacts.entities.custom.gender.*

class IntegrateGenderCustomDataActivity : Activity() {

    val contacts = Contacts(this, false, CustomDataRegistry().register(GenderRegistration()))

    fun getContactsWithGenderCustomData(): List<Contact> = contacts
        .query()
        .where { GenderFields.Type.isNotNull() }
        .find()

    fun insertRawContactWithGenderCustomData(): Insert.Result = contacts
        .insert()
        .rawContact {
            setGender(contacts) {
                type = GenderEntity.Type.MALE
            }
        }
        .commit()

    fun updateRawContactGenderCustomData(rawContact: RawContact): Update.Result = contacts
        .update()
        .rawContacts(
            rawContact.mutableCopy {
                gender(contacts)?.type = GenderEntity.Type.FEMALE
            }
        )
        .commit()

    fun deleteGenderCustomDataFromRawContact(rawContact: RawContact): Update.Result =
        contacts
            .update()
            .rawContacts(
                rawContact.mutableCopy {
                    setGender(contacts, null)
                }
            )
            .commit()

    fun getAllGender(): List<Gender> = contacts.data().query().genders().find()

    fun updateGender(gender: MutableGender): DataUpdate.Result =
        contacts.data().update().data(gender).commit()

    fun deleteGender(gender: Gender): DataDelete.Result =
        contacts.data().delete().data(gender).commit()
}